src/Pos/Generator/BlockEvent.hs

```diff
@@ -18,49 +19,201 @@ module Pos.Generator.BlockEvent
+buildBlockchainForest :: a -> Map Path a -> BlockchainForest a
+buildBlockchainForest defElem elements =
+    fmap (buildBlockchainTree defElem) . Map.fromListWith Map.union $ do
+        (Path path, a) <- Map.toList elements
+        case Seq.viewl path of
+            Seq.EmptyL -> error
+                "buildBlockchainForest: precondition violated, empty path"
+            pathSegment Seq.:< path' ->
+                [(pathSegment, Map.singleton (Path path') a)]
+
+buildBlockchainTree :: a -> Map Path a -> BlockchainTree a
+buildBlockchainTree defElem elements =
+    let
+        topPath = Path Seq.empty
+        topElement = Map.findWithDefault defElem topPath elements
+        -- 'otherElements' has its empty path deleted (if there was one in the
+        -- first place), so it satisfies the precondition of 'buildBlockchainForest'
+        otherElements = Map.delete topPath elements
+    in
+        BlockchainTree topElement (buildBlockchainForest defElem otherElements)
+
+-- Inverse to 'buildBlockchainForest'.
+flattenBlockchainForest' :: BlockchainForest a -> Map Path a
+flattenBlockchainForest' =
+    Map.fromList . flattenBlockchainForest (Path Seq.empty)
+
+flattenBlockchainForest :: Path -> BlockchainForest a -> [(Path, a)]
+flattenBlockchainForest (Path prePath) forest = do
+    (pathSegment, subtree) <- Map.toList forest
+    flattenBlockchainTree (Path $ prePath Seq.|> pathSegment) subtree
+
+flattenBlockchainTree :: Path -> BlockchainTree a -> [(Path, a)]
+flattenBlockchainTree prePath tree = do
+    let BlockchainTree a forest = tree
+    (prePath, a) : flattenBlockchainForest prePath forest
+
+type MonadGenBlockchain ctx m =
+    ( MonadBlockGen ctx m
+    , MonadThrow m )
+
+genBlocksInForest ::
+       (MonadGenBlockchain ctx m, RandomGen g)
+    => AllSecrets
+    -> BlockchainForest BlockDesc
+    -> RandT g m (BlockchainForest BlundDefault)
+genBlocksInForest secrets =
+    traverse $ mapRandT withClonedGState . genBlocksInTree secrets
+
+genBlocksInTree ::
+       (MonadGenBlockchain ctx m, RandomGen g)
+    => AllSecrets
+    -> BlockchainTree BlockDesc
+    -> RandT g m (BlockchainTree BlundDefault)
+genBlocksInTree secrets blockchainTree = do
+    let
+        BlockchainTree blockDesc blockchainForest = blockchainTree
+        txGenParams = case blockDesc of
+            BlockDescDefault  -> def
+            BlockDescCustom p -> p
+        blockGenParams = BlockGenParams
+            { _bgpSecrets     = secrets
+            , _bgpBlockCount  = 1
+            , _bgpTxGenParams = txGenParams
+            , _bgpInplaceDB   = True
+            }
+    -- Partial pattern-matching is safe because we specify
+    -- blockCount = 1 in the generation parameters.
+    OldestFirst [block] <- genBlocks blockGenParams
+    forestBlocks <- genBlocksInForest secrets blockchainForest
+    return $ BlockchainTree block forestBlocks
+
+-- Precondition: paths in the structure are non-empty.
+genBlocksInStructure ::
```

Comment:
I think it would be helpful to write some types explicitly. For instance, the type of `paths`.